producer/src/block/block.go

```diff
@@ -13,24 +13,37 @@ type Block struct {
 	PreviousHash   []byte
 	MerkleRootHash []byte
 	Data           [][]byte
+	DataLen        uint16
 }
 
 // Produces a block based on the struct provided
-func (b *Block) Serialize() []byte {
-	// allocates space for the known variables
-	serializedBlock := make([]byte, 20)
+func (b *Block) Serialize() []byte { // Vineet
+	//calculate the total length beforehand, to prevent unneccessary appends
+	//NOTE: 32 bit ints are used to hold lengths; unsigned 16 bit int is used for the length of Data
+	bLen := 4 + 8 + 8 + 32 + 32 + 2 //size of all fixed size fields
+	for _, s := range b.Data {
+		bLen += 4 + len(s) //4 bytes for length plus the length of an element in Data
+	}
+	serializedBlock := make([]byte, bLen)
 
 	// convert the known variables to byte slices in little endian and add to slice
 	binary.LittleEndian.PutUint32(serializedBlock[0:4], b.Version)
 	binary.LittleEndian.PutUint64(serializedBlock[4:12], b.Height)
```

Comment:
I am considering removing the `Height` field, because it limits the rate and amount of blocks we can generate. But that'll be a discussion for later, just putting that idea out there so don't worry about it right now.